strawberryfields/io.py

```diff
@@ -244,6 +246,145 @@ def is_free_param(param):
+
+        >>> print(code_str)
+        import strawberryfields as sf
+        from strawberryfields import ops
+
+        prog = sf.Program(3)
+        eng = sf.Engine("fock", backend_options={"cutoff_dim": 5})
+
+        with prog.context as q:
+            ops.Sgate(2*np.pi/3, 0.0) | q[1]
+            ops.BSgate(0.6, 0.1) | (q[2], q[0])
+            ops.MeasureFock() | (q[0], q[1], q[2])
+
+        results = eng.run(prog)
+
+    Args:
+        prog (Program): the Strawberry Fields program
+        eng (Engine): The Strawberryfields engine. If ``None``, only the Program
+            parts will be in the resulting code-string.
+
+    Returns:
+        str: the Strawberry Fields code, for constructing the program, as a string
+    """
+    code_seq = ["import strawberryfields as sf", "from strawberryfields import ops\n"]
+
+    if prog.type == "tdm":
+        code_seq.append(f"prog = sf.TDMProgram(N={prog.N})")
+    else:
+        code_seq.append(f"prog = sf.Program({prog.num_subsystems})")
+
+    # check if an engine is supplied; if so, format and add backend/target
+    # along with backend options
+    if eng:
+        eng_type = eng.__class__.__name__
+        if eng_type == "RemoteEngine":
+            code_seq.append(f'eng = sf.RemoteEngine("{eng.target}")')
+        else:
+            if "cutoff_dim" in eng.backend_options:
+                formatting_str = (
+                    f'"{eng.backend_name}", backend_options='
+                    + f'{{"cutoff_dim": {eng.backend_options["cutoff_dim"]}}}'
+                )
+                code_seq.append(f"eng = sf.Engine({formatting_str})")
+            else:
+                code_seq.append(f'eng = sf.Engine("{eng.backend_name}")')
+
+    # check if program is of TDM type and format the context as appropriate
+    if prog.type == "tdm":
+        # if the context arrays contain pi-values, factor out multiples of np.pi
+        tdm_params = [f"[{_factor_out_pi(par)}]" for par in prog.tdm_params]
+        code_seq.append("\nwith prog.context(" + ", ".join(tdm_params) + ") as (p, q):")
+    else:
+        code_seq.append("\nwith prog.context as q:")
+
+    # add the operations, and replace any free parameters with e.g. `p[0]`, `p[1]`
+    for cmd in prog.circuit:
+        name = cmd.op.__class__.__name__
+        if prog.type == "tdm":
+            format_dict = {k: f"p[{k[1:]}]" for k in prog.parameters.keys()}
+            params_str = _factor_out_pi(cmd.op.p).format(**format_dict)
+        else:
+            params_str = _factor_out_pi(cmd.op.p)
+
+        modes = [f"q[{r.ind}]" for r in cmd.reg]
+        if len(modes) == 1:
+            modes_str = ", ".join(modes)
+        else:
+            modes_str = "(" + ", ".join(modes) + ")"
+        op = f"    ops.{name}({params_str}) | {modes_str}"
+
+        code_seq.append(op)
+
+    if eng:
+        code_seq.append("\nresults = eng.run(prog)")
+
+    return "\n".join(code_seq)
+
+
+def _factor_out_pi(num_list, precision=12):
```

Comment:
Wondering if it would be worth adjusting `precision` to the value that can be set by [`np.set_printoptions`](https://numpy.org/doc/stable/reference/generated/numpy.set_printoptions.html). Right now we'd always use `precision=12` internally, correct?

Reply:
This is a bit different though, and again, this might be a misleading argument name. Here `precision` simply means how large denominators it should use for the factors (e.g. `precision=12` means it tries to factor out `np.pi/12`, while `precision=2` means that `np.pi/2` works, but `np.pi/3` will return `1.0471975512`).

Reply:
Got it! Could maybe go with the name `denominator` or `denom`?